aten/src/ATen/native/UpSample.h

```diff
@@ -0,0 +1,249 @@
+#include <math.h>
+
+#include <ATen/ATen.h>
+
+namespace at {
+namespace native {
+
+static inline void check_dim_size(
+    const Tensor& data,
+    int64_t dim,
+    int64_t dim_size,
+    int64_t size) {
+  /* Check dimension size of a tensor */
+  AT_CHECK(
+      data.dim() == dim && data.size(dim_size) == size,
+      "Expected tensor of dimension ",
+      dim,
+      " and tensor.size[",
+      dim_size,
+      "] == ",
+      size,
+      " but got: dimension ",
+      data.dim(),
+      " and tensor.size[",
+      dim_size,
+      "] = ",
+      data.size(dim_size));
+}
+
+static inline void upsample_1d_shape_check(
+    const Tensor& data,
+    int64_t type_check,
+    int64_t nbatch,
+    int64_t nchannels,
+    int64_t input_width,
+    int64_t output_width) {
+  AT_CHECK(
+      input_width > 0 && output_width > 0,
+      "Input and output sizes should be greater than 0, but got input (W: ",
+      input_width,
+      ") and output (W: ",
+      output_width,
+      ")");
+
+  if (type_check == 0) {
+    AT_CHECK(
+        !data.numel() == 0 && data.dim() == 3,
```

Comment:
This is wrong. ! has higher precedence than ==.

Reply:
Actually, I guess this accidentally works because ! on an integer will convert 0 to 1 and non-zero to 0 and that happens to work out. But I hate it. Do `data.numel() != 0`, it's more idiomatic and less likely to give the reader hives.

Reply:
you're totally right! thanks for catch up that.